service/history/nDCEventsReapplication.go

```diff
@@ -0,0 +1,262 @@
+	"github.com/uber/cadence/common/clock"
+	"github.com/uber/cadence/common/cluster"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/metrics"
+)
+
+type (
+	nDCEventsReapplication struct {
+		timeSource      clock.TimeSource
+		historyCache    *historyCache
+		domainCache     cache.DomainCache
+		clusterMetadata cluster.Metadata
+		resetor         workflowResetor
+		metricsClient   metrics.Client
+		logger          log.Logger
+	}
+)
+
+func newNDCEventsReapplication(
+	timeSource clock.TimeSource,
+	resetor workflowResetor,
+	historyCache *historyCache,
+	domainCache cache.DomainCache,
+	metricsClient metrics.Client,
+	logger log.Logger,
+	clusterMetadata cluster.Metadata,
+) *nDCEventsReapplication {
+
+	return &nDCEventsReapplication{
+		timeSource:      timeSource,
+		historyCache:    historyCache,
+		domainCache:     domainCache,
+		metricsClient:   metricsClient,
+		logger:          logger,
+		clusterMetadata: clusterMetadata,
+		resetor:         resetor,
+	}
+}
+func (r *nDCEventsReapplication) reapplyEvents(
+	ctx ctx.Context,
+	domainID string,
+	workflowID string,
+	historyEvents []*shared.HistoryEvent,
+) (retError error) {
+	reapplyEvents := []*shared.HistoryEvent{}
+	for _, event := range historyEvents {
+		switch event.GetEventType() {
+		case shared.EventTypeWorkflowExecutionSignaled:
+			reapplyEvents = append(reapplyEvents, event)
+		}
+	}
+
+	if len(reapplyEvents) == 0 {
+		return nil
+	}
+	context, msBuilder, releaseFunc, err := r.getCurrentWorkflowMutableState(ctx, domainID, workflowID)
+	defer func() { releaseFunc(retError) }()
+	if err != nil {
+		// for get workflow execution context, with valid run id
+		// err will not be of type EntityNotExistsError
+		return err
+	}
+
+	if msBuilder.IsWorkflowExecutionRunning() {
+		return r.reapplyEventsToCurrentRunningWorkflow(ctx, context, msBuilder, reapplyEvents)
+	}
+	return r.reapplyEventsToCurrentClosedWorkflow(ctx, context, msBuilder, reapplyEvents)
+}
+
+func (r *nDCEventsReapplication) reapplyEventsToCurrentRunningWorkflow(
+	ctx ctx.Context,
+	context workflowExecutionContext,
+	msBuilder mutableState,
+	events []*shared.HistoryEvent,
+) error {
+
+	canMutateWorkflow, err := r.prepareWorkflowMutation(msBuilder)
+	if err != nil || !canMutateWorkflow {
+		return err
+	}
+
+	numSignals := 0
+	for _, event := range events {
+		switch event.GetEventType() {
+		case shared.EventTypeWorkflowExecutionSignaled:
+			attr := event.WorkflowExecutionSignaledEventAttributes
+			if _, err := msBuilder.AddWorkflowExecutionSignaled(
+				attr.GetSignalName(),
+				attr.Input,
+				attr.GetIdentity()); err != nil {
+				return ErrWorkflowMutationSignal
+			}
+			numSignals++
+
+		default:
+			return ErrUnreappliableEvent
+		}
+	}
+
+	r.logger.Info(fmt.Sprintf("reapplying %v signals", numSignals))
+	now := r.timeSource.Now()
+	return context.updateWorkflowExecutionAsActive(now)
+}
+
+func (r *nDCEventsReapplication) reapplyEventsToCurrentClosedWorkflow(
```

Comment:
Please leave out the reapply to closed workflow for now, we either need to rewrite the existing code, supporting #2420 or introduce this back when reapply policy is introduced

Reply:
ignore the closed workflow for now. Added a todo here